File: chroma-league/src/main/java/com/bonepl/chromaleague/rest/gamestats/MapTerrain.java
package com.bonepl.chromaleague.rest.gamestats;

import com.bonepl.chromaleague.rest.eventdata.DragonType;
import com.bonepl.razersdk.color.Color;
import com.bonepl.razersdk.color.StaticColor;

import java.util.Arrays;
import java.util.NoSuchElementException;

public enum MapTerrain {
    CHEMTECH("Chemtech", DragonType.CHEMTECH),
    CLOUD("Cloud", DragonType.CLOUD),
    HEXTECH("Hextech", DragonType.HEXTECH),
    INFERNAL("Infernal", DragonType.INFERNAL),
    MOUNTAIN("Mountain", DragonType.MOUNTAIN),
    OCEAN("Ocean", DragonType.OCEAN);

    private final String apiType;
    private final DragonType dragonType;

    MapTerrain(String apiType, DragonType dragonType) {
        this.apiType = apiType;
        this.dragonType = dragonType;
    }

    public static MapTerrain fromApiType(String apiType) {
        return Arrays.stream(values())
                .filter(mt -> mt.apiType.equals(apiType))
                .findAny()
                .orElseThrow(() -> new NoSuchElementException("MapTerrain for apiType " + apiType + " does not exist"));
    }

    public Color getBackgroundColor() {
        StaticColor color = dragonType.getColor();
        return new StaticColor(reduceColor(color.red()), reduceColor(color.green()), reduceColor(color.blue()));
    }

    public int reduceColor(int color) {
        return (int) (0.1 * color);
    }
}

File: src/rest/gamestats/MapTerrain.ts
import {Color} from '../../razer-sdk/color/Color.js';
import {StaticColor} from '../../razer-sdk/color/StaticColor.js';
import {DragonType, dragonTypeGetColor} from '../eventdata/DragonType.js';

export enum MapTerrain {
    CHEMTECH = 'CHEMTECH',
    CLOUD = 'CLOUD',
    HEXTECH = 'HEXTECH',
    INFERNAL = 'INFERNAL',
    MOUNTAIN = 'MOUNTAIN',
    OCEAN = 'OCEAN',
}

const API_TYPE_MAP: Record<string, MapTerrain> = {
    Chemtech: MapTerrain.CHEMTECH,
    Cloud: MapTerrain.CLOUD,
    Hextech: MapTerrain.HEXTECH,
    Infernal: MapTerrain.INFERNAL,
    Mountain: MapTerrain.MOUNTAIN,
    Ocean: MapTerrain.OCEAN,
};

const DRAGON_TYPE_MAP: Record<MapTerrain, DragonType> = {
    [MapTerrain.CHEMTECH]: DragonType.CHEMTECH,
    [MapTerrain.CLOUD]: DragonType.CLOUD,
    [MapTerrain.HEXTECH]: DragonType.HEXTECH,
    [MapTerrain.INFERNAL]: DragonType.INFERNAL,
    [MapTerrain.MOUNTAIN]: DragonType.MOUNTAIN,
    [MapTerrain.OCEAN]: DragonType.OCEAN,
};

export function mapTerrainFromApiType(apiType: string): MapTerrain {
    const result = API_TYPE_MAP[apiType];
    if (result === undefined) {
        throw new Error(`MapTerrain for apiType ${apiType} does not exist`);
    }
    return result;
}

export function mapTerrainDragonType(terrain: MapTerrain): DragonType {
    return DRAGON_TYPE_MAP[terrain];
}

export function mapTerrainGetBackgroundColor(terrain: MapTerrain): Color {
    const color = dragonTypeGetColor(mapTerrainDragonType(terrain));
    return new StaticColor(
        Math.trunc(0.06 * color.red),
        Math.trunc(0.06 * color.green),
        Math.trunc(0.06 * color.blue),
    );
}
